src/Microsoft.Health.Checkpoints/Storage/BaseStorageCheckpointClient.cs

```diff
@@ -0,0 +1,90 @@
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure;
+using Azure.Storage.Blobs;
+using Azure.Storage.Blobs.Models;
+using EnsureThat;
+
+namespace Microsoft.Health.Checkpoints.Storage
+{
+    public class BaseStorageCheckpointClient : ICheckpointClient
+    {
+        private BlobContainerClient _storageClient;
+
+        public BaseStorageCheckpointClient(BlobContainerClient containerClient)
+        {
+            EnsureArg.IsNotNull(containerClient);
+
+            _storageClient = containerClient;
+        }
+
+        public virtual Task<ICheckpoint> GetCheckpointAsync(string checkpointIdentifier)
+        {
+            Task<ICheckpoint> GetCheckpointAsync()
+            {
+                ICheckpoint checkpoint = new Checkpoint();
+
+                foreach (BlobItem blob in _storageClient.GetBlobs(traits: BlobTraits.Metadata, states: BlobStates.All, prefix: checkpointIdentifier, cancellationToken: CancellationToken.None))
+                {
+                    DateTimeOffset lastEventTimestamp = DateTime.MinValue;
+
+                    if (blob.Metadata.TryGetValue("LastProcessedDt", out var dt))
+                    {
+                        DateTimeOffset.TryParse(dt, null, DateTimeStyles.AssumeUniversal, out lastEventTimestamp);
+                        checkpoint.LastProcessedDt = lastEventTimestamp;
+                    }
+
+                    if (blob.Metadata.TryGetValue("LastProcessedId", out var id))
+                    {
+                        checkpoint.LastProcessedId = id;
+                    }
+
+                    checkpoint.Identifier = checkpointIdentifier;
+                }
+
+                return Task.FromResult(checkpoint);
+            }
+
+            return GetCheckpointAsync();
+        }
+
+        public virtual async Task SetCheckpointAsync(ICheckpoint checkpoint)
+        {
+            EnsureArg.IsNotNull(checkpoint);
+            EnsureArg.IsNotNullOrWhiteSpace(checkpoint.Identifier);
+
+            var lastProcessedDt = EnsureArg.IsNotNullOrWhiteSpace(checkpoint.LastProcessedDt.DateTime.ToString("MM/dd/yyyy hh:mm:ss.fff tt"), nameof(checkpoint.LastProcessedDt));
+            var lastProcessedId = checkpoint.LastProcessedId;
+
+            var blobName = checkpoint.Identifier;
+            var blobClient = _storageClient.GetBlobClient(blobName);
+
+            var metadata = new Dictionary<string, string>()
+            {
+                { "LastProcessedDt",  lastProcessedDt },
+                { "LastProcessedId",  lastProcessedId },
+            };
+
+            try
+            {
+                await blobClient.SetMetadataAsync(metadata);
+            }
+            catch (RequestFailedException ex) when ((ex.ErrorCode == BlobErrorCode.BlobNotFound) || (ex.ErrorCode == BlobErrorCode.ContainerNotFound))
```

Comment:
Interested in knowing the thinking behind not using ExistsAsync https://docs.microsoft.com/en-us/dotnet/api/azure.storage.blobs.blobcontainerclient.existsasync.  Is the thought to optimized for the normal scenario where the checkpoint exists and rather than always doing two blob client calls (Exists & SetMetadata)?

Reply:
It seemed better than always making two blob client calls, but the real reason is that this code was shamelessly "borrowed" from the event hubs checkpointing code

Reply:
That is fine and makes sense.  One suggestion, if we are already using Polly elsewhere in the Shared repo it might be worth leveraging that library here for a slightly cleaner implementation.  Look at line 70 in https://github.com/microsoft/iomt-fhir/blob/master/src/lib/Microsoft.Health.Fhir.R4.Ingest/Service/R4FhirImportService.cs for an example.  Here we handle an optimistic concurrency failure with a refresh of the data from the server and a fresh merge with the internal data.